service/history/workflow/update/update.go

```diff
@@ -27,84 +27,358 @@ package update
-	u.state = stateAccepted
+// WaitOutcome observes this Update's acceptance, returning a Failure if the
+// update is rejected or nil if it is successfully accepted. This call will
+// block until the Outcome is known or the provided context.Context expires.
+func (u *Update) WaitAccepted(ctx context.Context) (*failurepb.Failure, error) {
+	return u.accepted.Get(ctx)
 }
 
-func (u *Update) complete(o *updatepb.Outcome) {
-	u.state = stateCompleted
-	u.pendingOutcome = o
+// OnMessage delivers a message to the Update state machine. The proto.Message
+// parameter is expected to be one of *updatepb.Request, *updatepb.Response,
+// *updatepb.Rejection, *updatepb.Acceptance, or a *protocolpb.Message whose
+// Body field contains an instance from the same list. Writes to the EventStore
+// occur synchronously but externally observable effects on this Update (e.g.
+// emmitting an Outcome or an Accepted) are registered with the EventStore to be
+// applied after the durable updates are committed. If the EventStore rolls
+// back its effects, this state machine does the same.
+func (u *Update) OnMessage(
+	ctx context.Context,
+	msg proto.Message,
+	eventStore EventStore,
+) error {
+	if msg == nil {
+		return invalidArgf("Update %q received nil message", u.id)
+	}
+	if protocolMsg, ok := msg.(*protocolpb.Message); ok {
+		var dynbody types.DynamicAny
+		if err := types.UnmarshalAny(protocolMsg.Body, &dynbody); err != nil {
+			return err
+		}
+		msg = dynbody.Message
+	}
+	switch body := msg.(type) {
+	case *updatepb.Request:
+		return u.onRequestMsg(ctx, body, eventStore)
+	case *updatepb.Acceptance:
+		return u.onAcceptanceMsg(ctx, body, eventStore)
+	case *updatepb.Rejection:
+		return u.onRejectionMsg(ctx, body, eventStore)
+	case *updatepb.Response:
+		return u.onResponseMsg(ctx, body, eventStore)
+	default:
+		return invalidArgf("Message type %T not supported", body)
+	}
 }
 
-func (u *Update) reject(f *failurepb.Failure) {
-	u.state = stateRejected
-	u.pendingOutcome = &updatepb.Outcome{
-		Value: &updatepb.Outcome_Failure{
-			Failure: f,
-		},
+// PollOutboundMessages loads any oubound messages from this Update state
+// machine into the output slice provided.
+func (u *Update) PollOutboundMessages(out *[]*protocolpb.Message) {
+	if u.state != stateRequested {
+		// Update only sends messages to the workflow when it is in
+		// stateRequested
+		return
 	}
+	*out = append(*out, u.request)
 }
 
-func (u *Update) notify() {
-	if u.pendingOutcome != nil {
-		u.outcome.Set(u.pendingOutcome, nil)
-		u.pendingOutcome = nil
+// onRequestMsg works if the Update is in any state but if the state is anything
+// other than stateAdmitted then it just early returns a nil error. This
+// effectively gives us update request deduplication by update ID. If the Update
+// is in stateAdmitted then it builds a protocolpb.Message that will be sent on
+// ensuing calls to PollOutboundMessages until the update is accepted.
+func (u *Update) onRequestMsg(
```

Comment:
The pattern for all the on\<Type\>Msg functions:

- assert expected state(s)
- validate message
- update msg counter
- write event(s) if applicable
- transition to next provisional state
- register OnAfterCommit to transition from Provisional to non-Provisional state and set Future(s) if applicable
- register OnAfterRollback to undo transition into provisional